{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/bergeric/miniconda3/lib/python3.5/site-packages/gffutils/interface.py:161: UserWarning: It appears that this database has not had the ANALYZE sqlite3 command run on it. Doing so can dramatically speed up queries, and is done by default for databases created with gffutils >0.8.7.1 (this database was created with version 0.8.7.1) Consider calling the analyze() method of this object.\n",
      "  \"method of this object.\" % self.version)\n"
     ]
    }
   ],
   "source": [
    "db = gffutils.FeatureDB('/Volumes/Promise_Pegasus/LCDB/lcdb-references/dmel/r6-11/gtf/dmel_r6-11.gtf.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3UTR\n",
      "5UTR\n",
      "CDS\n",
      "exon\n",
      "gene\n",
      "mRNA\n",
      "miRNA\n",
      "ncRNA\n",
      "pre_miRNA\n",
      "pseudogene\n",
      "rRNA\n",
      "snRNA\n",
      "snoRNA\n",
      "start_codon\n",
      "stop_codon\n",
      "tRNA\n",
      "transcript\n"
     ]
    }
   ],
   "source": [
    "for a in db.featuretypes():\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrX\tFlyBase\tgene\t19961297\t19969323\t.\t+\t.\tgene_id \"FBgn0031081\"; gene_symbol \"Nep3\";\n"
     ]
    }
   ],
   "source": [
    "for y in db.features_of_type('gene'):\n",
    "    print(y)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of transcripts per gene\n",
    "table1 = []\n",
    "for gene in db.features_of_type('gene'): \n",
    "    name = gene.id\n",
    "    transcript = db.children(gene, featuretype='transcript')\n",
    "    transcriptnum = len(list(transcript))\n",
    "    row = [name, transcriptnum]\n",
    "    table1.append(row)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>#transcripts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FBgn0031081</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBgn0052826</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBgn0031085</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBgn0062565</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBgn0031088</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene  #transcripts\n",
       "0  FBgn0031081             3\n",
       "1  FBgn0052826             1\n",
       "2  FBgn0031085             1\n",
       "3  FBgn0062565             1\n",
       "4  FBgn0031088             2"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame(table1, columns=['gene','#transcripts'])\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x110722a58>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAAEKCAYAAAAl5S8KAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFWpJREFUeJzt3XuwXeV53/Hvs2/nIlnoAiSmIGQhI4wdB1w1IYWhMXJs\nY1O3Q/oHaW2TSSaETgDj1nHiydDB4yYdp7FT4+m0sWNCPE1bt9h1PNDWTWJTd9q6sTBggzHyBWEj\nBLqAQDr3vffbP9Y60pY4ktBtvy/s72fmzNln781ZD0fox3Oe9a53RUoJSVJ+jdwFSJIqBrIkFcJA\nlqRCGMiSVAgDWZIKYSBLUiEMZEkqhIEsSYUwkCWpEK3jefOZZ56Z1q1bd5pKkaRXpvvvv393Sums\nY73vuAJ53bp1bNmy5cSrkqQRFBFPvJT3ObKQpEIYyJJUCANZkgphIEtSIQxkSSqEgSxJhTCQJakQ\nBrIkFcJAlqRCGMiSVAgDWZIKYSBLUiEMZEkqhIEsSYUwkCWpEAayJBXCQJakQhjIklQIA1mSCpE9\nkGdnZ3OXIElFyBrIO3fu5J3vfCePPPJIzjIkqQhZA3nXrl30ej22b9+eswxJKkLWQJ6bmwNgeno6\nZxmSVAQDWZIKUUQgz8zM5CxDkopQRCDbIUtS5kBeXPJmhyxJmQN5fn4esEOWJHBkIUnFKGJkMWUg\nS1IZI4upKQNZkorokB1ZSFIhM2RXWUhSIYE8O2OHLElFBPLCwgLdbjdnKZKUXRGBDI4tJKmIk3rg\niT1JyhzIBztkA1nSqMsayDNzs6RGu3rsyELSiMt7YcjcHKk9DtghS1L2k3qpPQHYIUtS5g55nn7L\nDlmSIGMgp5RYWJg/0CEbyJJGXbZAnp+fJ6VkIEtSLWsgA6TmGEQ4Q5Y08rIF8oGLQhpNotkxkCWN\nvGyBvHjZdGq0oNV2ZCFp5GUPZBpNUsNAlqTsgZwaLfqNloEsaeRlD2TqQPa+epJGXfZATvXIwvvq\nSRp12QOZRguazpAlKXsgL3bILnuTNOqyBzKNFqnZZnbWQJY02rIHcqpHFl3vqydpxGUP5MV1yOB+\nFpJGW/5AjiapaSBLUta9LKLZhghotgADWdJoy7vbW6MJ4H31JIncu701qs6YpoEsSVlnyKkOZGfI\nkpT7jiGHjSwMZEmjLOvIoh9VIGOHLEkZA3lujlQHcnKGLEkZA3lm4KReNCHCDlnSSMvbIdczZCKI\nVsdAljTSylj2BtB0xzdJoy3zKouDgex99SSNuoyBPHfgSj2AfqNlhyxppGUJ5JQS8wMXhgD0G22m\npqZylCNJRcgSyN1ul5TSIR1yarTY7331JI2wLIE8OzsLcEiH7H31JI26LIE8Pz9fH33gpF6zzcyM\ngSxpdGXukAdHFu0Dz0vSKMoSyIM3OD2g2abX7bKwsJCjJEnKLuvI4pAO2Q2GJI24rCMLDrswBAxk\nSaMr68ji8FUW4I5vkkZX5hmyIwtJWlROh+zIQtKIK2aVhZvUSxp1mTvkF48s3M9C0qjK2yHHQIfc\nXgaNJk888USOkiQpu2JO6tFo0J9YzWOPPZajJEnKLlsgR6MFEYc8351cw2Nbt1Y7wUnSiMl3YUiz\n9aLn+8vWMDM9zfbt2zNUJUl55bt0enBcUetNngnA1q1bh12SJGWXrUM+ZA1yrT+xEhpNA1nSSMo2\nQ05LdMg0mvQnVnliT9JIyjaySLFEIOOJPUmjK0sgz8zO0o8XjywA+svOZHpqiqeeemrIVUlSXvlW\nWSw1sgB6k2sAT+xJGj1FndQD6E+s8sSepJGU6aTe0sveAE/sSRpZGVdZLN0hQ3Vi77uPeWJP0mjJ\ns8pibu7QG5wepj+5hump/ezYsWOIVUlSXnkCeWF+6XXItd6y6oq9hx9+eFglSVJ2Qw/kbrdLv9c7\neoc8sQrGlvPRj36Uz372sywsLAyxQknKY+iBvNTm9C/SaLL/dX+X2TPWcuedd/JrN/w6DzzwgDNl\nSa9oQw/k2dnZ+shH7pABUnuC2QvezPSGt7DtqWd4//vfz6/feCP33XcfvV5vCJVK0nANPZDn5+cB\njrrKYlBv1Vr2veEXmT3/b7P1iR3cfvvtfOQjH7FblvSKk21kccR1yEtptFg4+yL2veFa5s65hPvu\nu49777339BQoSZlkG1m81A75ENFg/pxL6a04h0/ccQfbtm07tcVJUkZDD+T9+/fXRz6BQAaIYOY1\nV9Klye23f/hgxy1JL3NDDeTdu3fzBx/7GNEaqzajP0GpM8nU+Vewbdvj3HbbbXbKkl4R4nhOjm3a\ntClt2bLlhA60d+9ebrr5FrY/tYP9F76N9rM/pLHvaaI7T2p1oNk55P39ydXMrb3sqN+z/fQjTDz1\nTVK/y+arrmLz5s00m00igvn5efbu3ctzzz1Ho9HgwgsvZOPGjaxYseKE6pekExUR96eUNh3zfcMI\n5OnpaX7jppt4/PHH6Y2vIrU6NKd2MdFpc80113DPPfccXA5XS802vck1xwzmWJil/fS3Gd/1KKnX\nPWYtq9ecydlnncWaNatZuXIl4+PjdDodOp0O3W6XbrdLr9ej1WrRbrdpt9t0Oh3GxsYOvK/RGN4v\nFlHfmdtVJVI+y5cvZ9OmTTSbx7EYYcBLDeRjDnIj4gbgBoC1a9eeUDH3338/j//wh3SXnQ2LYZYS\n11xzDTfddBMpJT7/+c+f0PdO7XHmz/tbLPzkT9GYeRZSgtSHZpt+ZzmpPQ79Hq29P6az87s8u2cn\nz+7ZfULHkjS6brnlFq699trTeoxjBnJK6VPAp6DqkE/kIK9//etZtXo1e/fvZ9/Gq0njZzDxyBe5\n5557SCktuYStN7mGmYvecexvnvq09vyAiR0PwewLh7wUzTa0xyH1SXNTB56fmJxk1arVrF61iomJ\nqkNut9v0er0DXfJidzzMblh6OYmIV+Rvbkv9e51zzjm89a1vPf3HHtYM+YknnuCmm29h31yP/Ruv\npvPMIyc9Q27u/RGTT/41zLzAhg2v5S1vOThDnpub4/nnn3/RDHn9+vVMTk6e0L+DJJ2IUzayOFXO\nP/98/vDjH+N977sVtn6Zfa97F7Q6x/4Hj6AxtZvJH3yFteedxw2/9kEuv/zyA/NWSXo5Gurv4xs2\nbOC3fuuDMPsCzemTmOP2Flj2+P9k9apVfPKOO7jiiisMY0kve0PrkBedffbZ1YP+iW8QNP6jr8Ps\nC9z2ex/njDPOOEWVSVJeQz9j1elUY4roH3uJ2lJau79Pe/f3eM+7382ll156KkuTpKyG3iGPjY1V\nD44nkFOiue9pOk9/m9bzT/K6iy/m+uuvPz0FSlImQw/k8fFxAOIljiwa088x8aP/Q2PfM6xYcQb/\n4Fd+hWuvvZZWa+ilS9JpNfRUWxxZHLND7vfo7PgWY08/xKuWL+dXb72Vq6+++mCHLUmvMBk75KME\ncuqzfOt/I/btZPPmzdx8882sXHnimxFJ0svB0AO51WpVS9SOMrJozD5P7NvJjTfeyHXXXTfE6iQp\nn6GvsogIOmNjR+2QG1PVGuXLLjv6lXqS9EqSZaOGTmfsqB1yc2oPY2PjnHfeeUOsSpLyyhTInaN2\nyM2ZPWzYsOGEt7qTpJejLIE8Pj5+5FUWqU9reg8XXbRxuEVJUmZ5Anls7IjrkBuzz5N6XS688MIh\nVyVJeeUJ5Ikjd8iNqT0ABrKkkZOvQ05Ld8jN6d20Ox1P6EkaOdlmyI0jBnJ1Qs9LoyWNmoyrLJYI\n5NSnNf0sF230hJ6k0ZMlkMfGxoj04hlyY/YFUm/B+bGkkZRv2VtviUCur9DbaIcsaQRlG1ksdaVe\nc3oP7U6HtWvXZqhKkvLK1iGn3gIcdsfr5vRuNlxwgSf0JI2kbDNkAAZXWqREa/pZxxWSRlbeQB4Y\nW8T8FKm3wPr163OUJEnZZQ3kwQ2GorcAwIoVK3KUJEnZFdMh068CeWJiIkNFkpRfcR2ygSxpVGXu\nkF8cyJOTkzlKkqTsMnfILx5ZGMiSRpUdsiQVwhmyJBWiqFUWjWazuqxakkZQUR3y+Ng4EZGjJEnK\nLt9ub3DolXq9BSYmHVdIGl35dnvj0A6Z/oIn9CSNtHx3DIk4bJVFl2WTy3KUI0lFyBLIEUG7feht\nnKK/wLJldsiSRleWQAbojHUO6ZCb/a4jC0kjLVsgV/fVO7RDNpAljbKMHfLYIR0yvQUvCpE00rIF\n8sTY+CGrLFJ33kCWNNKyBfL4+NjBdcj9HqS+IwtJIy1jIA90yO70Jkl5T+o16pN67vQmSYWssnCn\nN0nKHciLM2QDWZLyBvLisrdwhixJeU/q0asDuf5sIEsaZfkuDOl0SL0upOQqC0kid4dMgtT3pJ4k\nkXuGDNDvuuxNksg8sgCqlRa9BZreT0/SiMs8sqDqkPsLjI07rpA02rKPLKLfI3oLTDo/ljTisgcy\n/W51Pz3vFiJpxGUP5KhP6i3zhJ6kEZc9kOl3afS7LFvmDU4ljbbsgRz9Hg1v3yRJ+QO5WmXR9aIQ\nSSMveyBH6kHP2zdJUvZAptcldR1ZSFL2C0OiN+/99CSJEi6d7s4A7mMhSdkCOSJodzrEwizgTm+S\nlC2QoeqSY8EOWZIgeyCPEd2qQzaQJY26rIE8Pj5Go+6QHVlIGnVZA3lsbKxaZYEdsiRlDeSJxT2R\nsUOWpMwji4OBbIcsadRlH1ksMpAljboiArnZatFut3OWIknZFRHIzo8lqZBAHvcGp5JURiB7+yZJ\nKmSVxTJvcCpJuS+drnZ88356klTIyMKTepJUyMjCNciSVMjIwkCWpEI6ZEcWklTIDNkOWZIKCWQ7\nZEkqJJDtkCUpcyCvXLkSgDVr1uQsQ5KK0Mp58HPPPZfPfOYzrF+/PmcZklSErIEMcMEFF+QuQZKK\nkHVkIUk6yECWpEIYyJJUCANZkgphIEtSIQxkSSqEgSxJhTCQJakQBrIkFcJAlqRCGMiSVAgDWZIK\nYSBLUiEMZEkqhIEsSYUwkCWpEAayJBXCQJakQhjIklSISCm99DdH7AKeOMljngnsPsnvcTpZ38mx\nvpNjfSen1PrOTymddaw3HVcgnwoRsSWltGmoBz0O1ndyrO/kWN/JKb2+Y3FkIUmFMJAlqRA5AvlT\nGY55PKzv5FjfybG+k1N6fUc19BmyJGlpjiwkqRBDDeSIeHtEPBYR34+I3x7msY9Qz50RsTMiHh54\nbnVE/EVEfK/+vCpjfedFxFcj4jsR8UhEvK+UGiNiPCL+OiIeqmv7cCm1HVZnMyIeiIh7SqsvIrZF\nxLcj4sGI2FJgfSsj4u6I+G5EPBoRP1dKfRGxsf65LX68EBG3llLfiRpaIEdEE/jXwNXAxcAvRcTF\nwzr+EdwFvP2w534b+KuU0muBv6q/zqUL/NOU0sXAZcBv1D+zEmqcA65KKf00cAnw9oi4rJDaBr0P\neHTg69Lqe3NK6ZKBpVol1fcJ4L+nlC4Cfprq51hEfSmlx+qf2yXA3wSmgf9SSn0nLKU0lA/g54Av\nD3z9IeBDwzr+UepaBzw88PVjwKvrx68GHstd40Btfw78Qmk1ApPAN4GfLak24Fyqv5RXAfeU9ucL\nbAPOPOy5IuoDzgAepz7PVFp9h9X0VuB/l1rf8XwMc2TxN4AfD3z9ZP1caX4ipbSjfvw08BM5i1kU\nEeuAS4H/RyE11uOAB4GdwF+klIqprfavgA8C/YHnSqovAX8ZEfdHxA31c6XU9xpgF/An9cjnjyNi\nWUH1DboO+A/14xLre8k8qXcUqfrfbPZlKBGxHPg8cGtK6YXB13LWmFLqpepXxnOBn4mIN5RSW0Rc\nA+xMKd1/pPcU8Od7Rf3zu5pqHHXl4IuZ62sBbwL+TUrpUmCKw379L+DnR0R0gHcB//nw10qo73gN\nM5C3A+cNfH1u/VxpnomIVwPUn3fmLCYi2lRh/GcppS/UTxdVY0ppL/BVqnl8KbVdDrwrIrYB/xG4\nKiL+XUH1kVLaXn/eSTX//JmC6nsSeLL+rQfgbqqALqW+RVcD30wpPVN/XVp9x2WYgfwN4LUR8Zr6\n/2rXAV8a4vFfqi8B19ePr6ea22YREQF8Bng0pfTxgZey1xgRZ0XEyvrxBNVs+7sl1AaQUvpQSunc\nlNI6qv/WvpJSencp9UXEsoh41eJjqjnow6XUl1J6GvhxRGysn9oMfIdC6hvwSxwcV0B59R2fIQ/f\n3wFsBX4A/E7uATrVH+QOYIGqI/hVYA3ViaDvAX8JrM5Y3xVUv3J9C3iw/nhHCTUCbwQeqGt7GPhn\n9fPZa1ui1p/n4Em9IuoD1gMP1R+PLP59KKW+upZLgC31n/EXgVWF1bcM2AOcMfBcMfWdyIdX6klS\nITypJ0mFMJAlqRAGsiQVwkCWpEIYyJJUCANZp01E/IuIeHNE/P2I+FD93C9HxDm5awOIiE0Rcccx\n3rMuIv7hsGrSaDOQdTr9LPB14O8AX6uf+2VgyUCudwQciohopZS2pJRuOcZb1wEGsobCdcg65SLi\nXwJvo9qg5gfABVQ7h91NtdnPdmCGagfAR4HPUV3p9/vAq4AbgA7wfeA9KaXpiLgLeAHYBPwk8MGU\n0t315bGfA1ZQ7b/wj1NK/ysi3g78HtAEdqeUNkfE7XUt64EfAX8EfCCldM3AaxuobiX/+ymlT0fE\n14HX1fX/KfA/gD+p62sAv5hS+t6p/hlqNLVyF6BXnpTSb0bEfwLeC/wT4L6U0uUAEXEVVQgubsgO\nsCel9Kb66zUppU/Xj/851dWTn6y/9auprl68iOoS2buputcvp5R+t+6wJyPiLODTwJUppccjYvVA\neRdTbeozExE/f1jpb6Tad3oZ8EBE3Eu1oc4HUkrX1DV9EvhESunP6i0AhtbV65XPQNbp8iaqy4Iv\n4tAN4pfyuYHHb6iDeCWwHPjywGtfTCn1ge9ExOK2it8A7qw3YfpiSunBOmi/llJ6HCCl9OzA9/hS\nSmnmCHX8ef3aTER8lWqzn72Hvef/Ar8TEecCX7A71qnkDFmnVERcUu+R/LvAB4B7gbfVt9mZOMI/\nNjXw+C7gppTSTwEfBsYHXpsbPBRASulrwJVUY5C7IuK9xyhx6iivHT6/e9E8L6X076m2e5wB/mvd\n8UunhIGsUyql9GCq9vjdSjUe+ArwtlTdbmcG2Ec1Jz6SVwE76o73Hx3reBFxPvBMPeb4Y6rO/OvA\nlRHxmvo9q4/yLQb9vajuFbiGakOibxxeb0SsB36YUrqDaiexN77E7y0dkyMLnXL1DPe5lFI/Ii5K\nKX1n4OW7gH8bEYsn9Q53G9VdUXbVn48W3lAF529GxAKwH3hvSmlXfQeOL0REg2pP3F94CaV/i2pf\n5zOBj6SUnoqIXUAvIh6qax8D3lMf72mqE4fSKeEqCwmoV1nsTyn9Qe5aNLocWUhSIeyQJakQdsiS\nVAgDWZIKYSBLUiEMZEkqhIEsSYUwkCWpEP8fHO23/C5BaaUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1068a4828>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sb.violinplot(df1['#transcripts'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "newdf = df1.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3UTR\n",
      "5UTR\n",
      "CDS\n",
      "exon\n",
      "gene\n",
      "mRNA\n",
      "miRNA\n",
      "ncRNA\n",
      "pre_miRNA\n",
      "pseudogene\n",
      "rRNA\n",
      "snRNA\n",
      "snoRNA\n",
      "start_codon\n",
      "stop_codon\n",
      "tRNA\n",
      "transcript\n"
     ]
    }
   ],
   "source": [
    "introns = list(db.create_introns(exon_featuretype='exon',grandparent_featuretype=None,parent_featuretype='transcript'))\n",
    "for a in db.featuretypes():\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrX\tFlyBase\tmRNA\t19961689\t19968479\t15\t+\t.\ttranscript_symbol \"Nep3-RA\"; transcript_id \"FBtr0070000\"; gene_id \"FBgn0031081\"; gene_symbol \"Nep3\";\n"
     ]
    }
   ],
   "source": [
    "for x in db.features_of_type('mRNA'):\n",
    "    print(x)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrX\tgffutils_derived\ttranscript\t19961689\t19968479\t.\t+\t.\ttranscript_id \"FBtr0070000\"; gene_id \"FBgn0031081\";\n"
     ]
    }
   ],
   "source": [
    "for x in db.features_of_type('transcript'): \n",
    "    if x.id=='FBtr0070000':\n",
    "        print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'FBtr0301831'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = []\n",
    "flag = False\n",
    "for trns in db.features_of_type('transcript'):\n",
    "    name = trns.id\n",
    "    exons = list(db.children(trns, featuretype='exon'))\n",
    "  \n",
    "    #ints = list(db.children(trns, featuretype='intron'))\n",
    "    # length of 1st exon for each transcript\n",
    "    firstex = exons[0]\n",
    "    exlen = firstex.end - firstex.start\n",
    "    # number of introns/exons\n",
    "    numex = len(exons)\n",
    "    numin = (numex - 1)\n",
    "    stuff = [name, exlen, numex, numin]\n",
    "    table.append(stuff) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>1st_ex_len</th>\n",
       "      <th>#ex</th>\n",
       "      <th>#int</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FBtr0081624</td>\n",
       "      <td>298</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBtr0071763</td>\n",
       "      <td>240</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBtr0071764</td>\n",
       "      <td>818</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBtr0100521</td>\n",
       "      <td>58</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBtr0342981</td>\n",
       "      <td>58</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    transcript  1st_ex_len  #ex  #int\n",
       "0  FBtr0081624         298    1     0\n",
       "1  FBtr0071763         240   10     9\n",
       "2  FBtr0071764         818    9     8\n",
       "3  FBtr0100521          58   10     9\n",
       "4  FBtr0342981          58    9     8"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(table, columns=['transcript','1st_ex_len','#ex','#int'])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tlist = df['transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# length of coding sequence \n",
    "table2 = []\n",
    "cdslen = 0\n",
    "for x in tlist:\n",
    "    if list(db.children(x, featuretype='CDS')): \n",
    "        for i in db.children(x, featuretype='CDS'):\n",
    "            len1 = (len(i) - 1 )\n",
    "            cdslen += len1\n",
    "    else:     \n",
    "        cdslen = 0\n",
    "    \n",
    "    newline = [x, cdslen]\n",
    "    table2.append(newline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# length of utrs using transcript\n",
    "table3 = []\n",
    "utrlen = 0\n",
    "for x in tlist: \n",
    "    if list(db.children(x, featuretype='3UTR')):\n",
    "        for utr in db.children(x, featuretype='3UTR'):\n",
    "            len2 = len(utr) - 1 \n",
    "            utrlen += len2\n",
    "    else:\n",
    "        utrlen = 0\n",
    "    newline = [x, utrlen]\n",
    "    table3.append(newline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "table4 = []\n",
    "utrlen = 0\n",
    "for x in tlist: \n",
    "    if list(db.children(x, featuretype='5UTR')):\n",
    "        for utr in db.children(x, featuretype='5UTR'):\n",
    "            len2 = len(utr) - 1 \n",
    "            utrlen += len2\n",
    "    else:\n",
    "        utrlen = 0\n",
    "    newline = [x, utrlen]\n",
    "    table4.append(newline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mrna utrs\n",
    "table4 = []\n",
    "utrln=0\n",
    "for x in db.features_of_type('mRNA'):\n",
    "    for utr in db.region(x, featuretype='3UTR'):\n",
    "        ln = len(utr) - 1\n",
    "        utrln += ln\n",
    "    newline = [x, utrln]\n",
    "    table4.append(newline)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "utr3frame = pd.DataFrame(table3, columns=['transcript','3UTR_len'])\n",
    "utr5frame = pd.DataFrame(table4, columns=['transcript','5UTR_len'])\n",
    "utr3frame.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = pd.DataFrame(table2, columns=['transcript','cds_len'])\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3['cds_len'].sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bigdf = df.merge(df3, how='left', on='transcript').merge(utr3frame, how='left', on='transcript').merge(utr5frame,how='left', on='transcript')\n",
    "bigdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat([bigdf.describe(), newdf], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
